mistralrs-core/src/pipeline/gguf_tokenizer.rs

```diff
@@ -19,115 +20,71 @@ pub struct ConversionResult {
-            let scores = scores
-                .as_ref()
-                .expect("Expect `tokenizer.ggml.scores` for `llama` unigram tokeizer.");
-            let mut vocab = Vec::new();
-            for (token, score) in tokens.iter().zip(scores) {
-                vocab.push((token.clone(), *score as f64));
-            }
+    let metadata = MetadataContext {
+        path_prefix: "tokenizer.ggml".to_string(),
+        metadata: &content.metadata,
+    };
+    let props = PropsGGUF::try_from(metadata)?;
 
-            // Unigram (sentencepiece) default UNK is 0
-            let unk = unk.map(|x| x as usize).unwrap_or(0);
-            unk_str = tokens[unk].clone();
-
-            let unigram = Unigram::from(vocab, Some(unk), true).map_err(anyhow::Error::msg)?;
-            let mut tokenizer = Tokenizer::new(ModelWrapper::Unigram(unigram));
-            tokenizer.with_decoder(decoders::sequence::Sequence::new(vec![
-                DecoderWrapper::Replace(Replace::new("▁", " ").map_err(anyhow::Error::msg)?),
-                DecoderWrapper::ByteFallback(ByteFallback::new()),
-                DecoderWrapper::Fuse(Fuse::new()),
-                DecoderWrapper::Strip(Strip::new(' ', 1, 0)),
-            ]));
-            tokenizer.with_normalizer(normalizers::Sequence::new(vec![
-                NormalizerWrapper::Prepend(Prepend::new("▁".to_string())),
-                NormalizerWrapper::Replace(Replace::new(" ", "▁").map_err(anyhow::Error::msg)?),
-            ]));
-
-            tokenizer.add_special_tokens(&[AddedToken::from(tokens[bos as usize].clone(), true)]);
-            tokenizer.add_special_tokens(&[AddedToken::from(tokens[eos as usize].clone(), true)]);
-            tokenizer.add_special_tokens(&[AddedToken::from(tokens[unk].clone(), true)]);
-
-            (tokenizer, "unigram")
-        }
+    let (tokenizer, kind, special_tokens) = match props.model.as_str() {
+        "llama" | "replit" => unigram_tokenizer(&props)?,
         other => {
             anyhow::bail!("Tokenizer model `{other}` not supported.");
         }
     };
+
     info!(
-        "GGUF tokenizer model is `{model}`, kind: `{}`, num tokens: {}, num added tokens: {}, num merges: {}, num scores: {}",
-        ty,
+        "GGUF tokenizer model is `{model}`, kind: `{kind:?}`, num tokens: {}, num added tokens: {}, num merges: {}, num scores: {}",
         tokenizer.get_vocab_size(true),
-        added_tokens.as_ref().map(|x| x.len()).unwrap_or(0),
-        merges.as_ref().map(|x| x.len()).unwrap_or(0),
-        scores.as_ref().map(|x| x.len()).unwrap_or(0)
+        props.added_tokens.as_ref().map(|x| x.len()).unwrap_or(0),
+        props.merges.as_ref().map(|x| x.len()).unwrap_or(0),
+        props.scores.as_ref().map(|x| x.len()).unwrap_or(0),
+        model = props.model,
     );
     if DEBUG.load(Ordering::Relaxed) {
         info!("Tokenizer: {tokenizer:?}");
     }
+
+    let [bos_str, eos_str, unk_str] = special_tokens
```

Comment:
> In the original code, the unk token is optional.

`unk` is still optional. I did shift some of the logic around but `unk_str` began as unset, relying upon `unigram` to set it, otherwise panic.

All I did was have the `special_tokens` returned from the supported tokenizer (eg: unigram) which made more sense to me since the final portion adds these tokens in the same manner you were setting them.

I've not changed the expectation though, below you assume `unk_str` is some value and set it as `Some(unk_str)`. If `unk_str` should be `Option<String>` instead, I can adjust for that.

Reply:
Ah, ok. Sounds good.